src/backend/db/database_manager.py

```diff
@@ -418,5 +422,436 @@ def delete_custom_field(self, customer_guid, field_name):
             logger.error(f"Error deleting custom field: {e}")
             session.rollback()
             return {"status": "error", "error": str(e)}
+        finally:
+            session.close()
+
+
+    # Tickets related function
+    def create_ticket(self, customer_guid, chat_id, title, description, priority, reported_by, assigned, custom_fields):
+        customer_db_name = self.get_customer_db(customer_guid)
+        session = DatabaseManager._session_factory()
+
+        try:
+            try:
+                # Check if the database exists
+                logger.debug(f"Checking if database {customer_db_name} exists")
+                result = session.execute(
+                    text("SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = :db_name"),
+                    {"db_name": customer_db_name}
+                ).fetchone()
+            except OperationalError as e:
+                logger.error(f"Database connectivity issue during schema check: {e}")
+                raise e
+
+            if not result:
+                raise ValueError(f"Database {customer_db_name} does not exist")
+
+            # Switch to the customer database
+            logger.debug(f"Switching to customer database: {customer_db_name}")
+            session.execute(text(f"USE `{customer_db_name}`"))
+
+            # Check if the chat_id exists in the chat_messages table
+            chat_exists = session.execute(
+                text("SELECT 1 FROM chat_messages WHERE chat_id = :chat_id LIMIT 1"),
+                {"chat_id": chat_id}
+            ).fetchone()
+
+            if not chat_exists:
+                raise ValueError(f"Invalid chat_id: {chat_id} does not exist.")
+
+            # Validate custom fields against the database columns
+            if custom_fields:
+                # Fetch the valid column names from the custom_field_values table
+                result = session.execute("SHOW COLUMNS FROM custom_field_values").fetchall()
+                valid_columns = {row[0] for row in result}
+
+                # Check for invalid fields
+                invalid_fields = [field for field in custom_fields.keys() if field not in valid_columns]
+                if invalid_fields:
+                    raise ValueError(f"Invalid custom field(s): {', '.join(invalid_fields)}")
+
+            # Begin transaction
+            with session.begin_nested():
+                # Insert into tickets table
+                query = '''INSERT INTO tickets (chat_id, title, description, priority, reported_by, assigned)
+                           VALUES (:chat_id, :title, :description, :priority, :reported_by, :assigned)'''
+                session.execute(
```

Comment:
This will be a huge concurrency problem which becomes hard to debug later

use the below code to circumvent this problem

Add a UUID column to the table to create uniqueness and then add the values and you can query it easily

import uuid

# Generate a unique identifier
ticket_uuid = str(uuid.uuid4())

# Insert with the unique ID
query = '''
INSERT INTO tickets (chat_id, title, description, priority, reported_by, assigned, ticket_uuid)
VALUES (:chat_id, :title, :description, :priority, :reported_by, :assigned, :ticket_uuid)
'''
session.execute(
    text(query),
    {
        "chat_id": chat_id,
        "title": title,
        "description": description,
        "priority": priority,
        "reported_by": reported_by,
        "assigned": assigned,
        "ticket_uuid": ticket_uuid
    },
)

# Fetch the row using the generated UUID
result = session.execute(
    text("SELECT * FROM tickets WHERE ticket_uuid = :ticket_uuid"),
    {"ticket_uuid": ticket_uuid}
).fetchone()

Reply:
modified sir